crates/rpc/src/rpc.rs

```diff
@@ -95,199 +55,66 @@ impl Rpc for Server {
     fn get_listings(&self) -> Result<Vec<Listing>> {
         let db = self.db()?;
 
-        // TODO: figure out a less ugly way to perform this join
-        let items: Vec<RpcGetListingsJoin> = listings::table
-            .inner_join(listing_metadatas::table.inner_join(metadatas::table))
-            .inner_join(storefronts::table)
-            .select((
-                listings::address,
-                listings::ends_at,
-                listings::created_at,
-                listings::ended,
-                listings::last_bid,
-                listings::price_floor,
-                listings::total_uncancelled_bids,
-                listings::instant_sale_price,
-                storefronts::subdomain,
-                storefronts::title,
-                metadatas::address,
-                metadatas::name,
-                metadatas::uri,
-            ))
-            .order_by((listings::address, listing_metadatas::metadata_index))
-            .load(&db)
-            .map_err(internal_error("Failed to load listings"))?;
-
-        let mut listings = HashMap::default();
-
-        for RpcGetListingsJoin {
-            address,
-            ends_at,
-            created_at,
-            ended,
-            last_bid,
-            price_floor,
-            total_uncancelled_bids,
-            instant_sale_price,
-            subdomain,
-            store_title,
-            meta_address,
-            name,
-            uri,
-        } in items
-        {
-            match listings.entry(address.clone()) {
-                Entry::Vacant(v) => {
-                    v.insert(Listing {
-                        address,
-                        ends_at: ends_at.map(|e| e.to_string()),
-                        created_at: created_at.to_string(),
-                        ended,
-                        last_bid,
-                        price_floor,
-                        total_uncancelled_bids,
-                        instant_sale_price,
-                        subdomain,
-                        store_title,
-                        items: vec![ListingItem {
-                            address: meta_address,
-                            name,
-                            uri,
-                        }],
-                    });
-                },
-                Entry::Occupied(o) => {
-                    let o = o.into_mut();
-                    o.items.push(ListingItem {
-                        address: meta_address,
-                        name,
-                        uri,
-                    });
-                },
-            }
-        }
-
-        Ok(listings.into_values().collect())
+        queries::listings_triple_join(|q| {
+            q.select(queries::LISTINGS_TRIPLE_JOIN_COLS)
```

Comment:
How do I access ::listings from this context? to apply a filter based on a listing f.ex